confluent_kafka/avro/serializer/message_serializer.py

```diff
@@ -151,33 +153,38 @@ def encode_record_with_schema_id(self, schema_id, record, is_key=False):
             return outf.getvalue()
 
     # Decoder support
-    def _get_decoder_func(self, schema_id, payload):
+    def _get_decoder_func(self, schema_id, payload, is_key=False):
         if schema_id in self.id_to_decoder_func:
             return self.id_to_decoder_func[schema_id]
 
-        # fetch from schema reg
+        # fetch writer schema from schema reg
         try:
-            schema = self.registry_client.get_by_id(schema_id)
+            writer_schema_obj = self.registry_client.get_by_id(schema_id)
         except ClientError as e:
             raise SerializerError("unable to fetch schema with id %d: %s" % (schema_id, str(e)))
 
-        if schema is None:
+        if writer_schema_obj is None:
             raise SerializerError("unable to fetch schema with id %d" % (schema_id))
 
         curr_pos = payload.tell()
+
+        reader_schema_obj = self.reader_key_schema if is_key else self.reader_value_schema
+
         if HAS_FAST:
             # try to use fast avro
             try:
-                schema_dict = schema.to_json()
-                schemaless_reader(payload, schema_dict)
+                writer_schema = writer_schema_obj.to_json()
+                reader_schema = reader_schema_obj.to_json()
```

Comment:
Sorry I'm a bit late to the party here.

@fkaufer , correct me if I'm wrong, but my understanding is that `self.reader_value_schema` can be null, and so `reader_schema_obj` could be null as well. Wouldn't that line cause an NPE?

Also I wonder if the next line should not be `schemaless_reader(payload, writer_schema, reader_schema)` to be consistent with the code executed by the lambda this function returns.

Reply:
The next line is a sanity check as alluded to in the comment below it. This is in place to catch edges which fastavro did not covet at the time the serializer was written. It's unclear if this is still actually an issue or not but it only happens upon instantiating the writer. 

You won't get an NPE as the `None` type is an actual null but you will get an attribute error which would cause the serializer to fall back to `slow avro` as we call it. If you have already done so please feel free to open an issue which I will address tomorrow.

Reply:
@rnpridgeon , thanks this is helpful. I had already opened an issue for a related issue: https://github.com/confluentinc/confluent-kafka-python/issues/518 I think both can be addressed at the same time. 

I actually tried to fix it my self, but couldn't get the test to run. I would need to spend more time trying to set up my development environment.